File: resilient/resilient-annotation-processor/src/main/java/io/koraframework/resilient/annotation/processor/aop/ResilientAopErrors.java
package io.koraframework.resilient.annotation.processor.aop;

import javax.lang.model.element.ExecutableElement;

final class ResilientAopErrors {
    private ResilientAopErrors() {}

    static String unsupportedReturnTypeError(String annotationName, ExecutableElement method, Object unsupportedType) {
        return """
            %s cannot be applied to '%s#%s()' because return type '%s' is not supported by this aspect.

            Fix: use a synchronous, CompletionStage-compatible, or supported reactive method for this aspect, depending on the Java AP capabilities.
            """.formatted(annotationName, method.getEnclosingElement(), method.getSimpleName(), unsupportedType).trim();
    }

    static String invalidResilientContractError(String annotationName, ExecutableElement method, String expectedContract) {
        return """
            %s on '%s#%s()' references an invalid resilient component type.

            The annotation value must extend %s.
            Fix: point the annotation value to a generated/spec interface or custom component that implements %s.
            """.formatted(annotationName, method.getEnclosingElement(), method.getSimpleName(), expectedContract, expectedContract).trim();
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/io/koraframework/resilient/symbol/processor/aop/ResilientAopErrors.kt
package io.koraframework.resilient.symbol.processor.aop

import com.google.devtools.ksp.symbol.KSFunctionDeclaration

internal fun unsupportedReturnTypeError(annotationName: String, function: KSFunctionDeclaration, unsupportedType: Any): String {
    return """
        $annotationName cannot be applied to '${function.parentDeclaration}#${function.simpleName.asString()}' because return type '$unsupportedType' is not supported by this aspect.

        Fix: use a synchronous, suspend, or Flow-returning method for this aspect, or choose an aspect implementation that supports this async type.
    """.trimIndent()
}

internal fun invalidResilientContractError(annotationName: String, function: KSFunctionDeclaration, expectedContract: String): String {
    return """
        $annotationName on '${function.parentDeclaration}#${function.simpleName.asString()}' references an invalid resilient component type.

        The annotation value must extend $expectedContract.
        Fix: point the annotation value to a generated/spec interface or custom component that implements $expectedContract.
    """.trimIndent()
}
